src/main/scala/mesosphere/marathon/core/volume/providers/DVDI.scala

```diff
@@ -0,0 +1,163 @@
+import scala.collection.JavaConverters._
+
+/**
+  * DVDIProvider (Docker Volume Driver Interface provider) handles persistent volumes allocated
+  * by a specific docker volume driver plugin. This works for both docker and mesos containerizers,
+  * albeit with some limitations:
+  *   - only a single volume driver per container is allowed when using the docker containerizer
+  *   - docker containerizer requires that referenced volumes be created prior to application launch
+  *   - mesos containerizer only supports volumes mounted in RW mode
+  */
+protected case object DVDIProvider extends InjectionHelper[PersistentVolume] with PersistentVolumeProvider {
+  import org.apache.mesos.Protos.Volume.Mode
+
+  override val name = Some("dvdi")
+
+  val optionDriver = name.get + "/driverName"
+  val optionIOPS = name.get + "/iops"
+  val optionType = name.get + "/volumeType"
+
+  val validOptions: Validator[Map[String, String]] = validator[Map[String, String]] { opt =>
+    opt.get(optionDriver) as "driverName option" is notEmpty
+    // TODO(jdef) stronger validation for contents of driver name
+    opt.get(optionDriver).each as "driverName option" is notEmpty
+    // TODO(jdef) validate contents of iops and volume type options
+  }
+
+  val validPersistentVolume = validator[PersistentVolume] { v =>
+    v.persistent.name is notEmpty
+    v.persistent.name.each is notEmpty
+    v.persistent.providerName is notEmpty
+    v.persistent.providerName.each is notEmpty
+    v.persistent.providerName.each is equalTo(name.get) // sanity check
+    v.persistent.options is valid(validOptions)
+  }
+
+  def nameOf(vol: PersistentVolumeInfo): Option[String] = {
+    if (vol.providerName.isDefined && vol.name.isDefined) {
+      Some(vol.providerName.get + "::" + vol.name.get)
+    }
+    else None
+  }
+
+  // group-level validation for DVDI volumes: the same volume name may only be referenced by a single
+  // task instance across the entire cluster.
+  val groupValidation: Validator[Group] = new Validator[Group] {
+    override def apply(g: Group): Result = {
+      val groupViolations = g.apps.flatMap { app =>
+        val nameCounts = volumeNameCounts(app)
+        val internalNameViolations = {
+          nameCounts.filter(_._2 > 1).map{ e =>
+            RuleViolation(app.id, s"Requested volume ${e._1} is declared more than once within app ${app.id}", None)
+          }
+        }
+        val instancesViolation: Option[RuleViolation] =
+          if (app.instances > 1) Some(RuleViolation(app.id,
+            s"Number of instances is limited to 1 when declaring external volumes in app ${app.id}", None))
+          else None
+        val ruleViolations = app.container.toSet[Container].flatMap(DVDIProvider.this.collect).flatMap{ vol =>
+          val name = nameOf(vol.persistent)
+          if (name.isDefined) {
+            for {
+              otherApp <- g.transitiveApps.toList
```

Comment:
This algorithm becomes quadratic in the number of apps. Better extend the `volumeNameCounts` mapping above to return lists of app ids, not the count. Then you can produce nice error messages in O(n*log(n)).